### PR TITLE
New incident management

### PR DESCRIPTION
**Pull request by kRhythm53**

### CODE REVIEW

File: handlers/incident.go [hunk 1]
```diff
@@ -33,6 +36,64 @@ func apiServiceIncidentsHandler(w http.ResponseWriter, r *http.Request) {
...
+func getVisibleIncidentsOfService(service *services.Service) []incidents.Incident {
+	var visibleIncidents []incidents.Incident
+	for _, incident := range service.Incidents {
+		if visibilityCheck(incident) == true {
+			incidentVar := *incident
+			reverse(incidentVar.Updates)
+			visibleIncidents = append(visibleIncidents, incidentVar)
+		}
+	}
+	log.Info(fmt.Sprintf("Visible Incidents of Service %v : %v", service.Name, visibleIncidents))
```
Comment (dmonark) [inline handlers/incident.go]: Don't print whole visibleIncidents. Only print ID

File: handlers/incident.go [hunk 2]
```diff
@@ -33,6 +36,64 @@ func apiServiceIncidentsHandler(w http.ResponseWriter, r *http.Request) {
...
+		incidents[i], incidents[j] = incidents[j], incidents[i]
+	}
+}
+
+func visibilityCheck(incident *incidents.Incident) bool {
+	incidentUpdates := incident.Updates
+	log.Infof(fmt.Sprintf("Latest Incident Update: %v, Time Diff: %v ", incidentUpdates[len(incidentUpdates)-1], timeDiff(incidentUpdates[len(incidentUpdates)-1])))
+	if len(incidentUpdates) == 0 || !(incidentUpdates[len(incidentUpdates)-1].Type == resolved && timeDiff(incidentUpdates[len(incidentUpdates)-1]) > incidentsTimeoutInMinutes) {
+		return true
+	}
```
Comment (dmonark) [inline handlers/incident.go]: Break this in 2 if conditions

**Review by dmonark — CHANGES_REQUESTED** [review]

**Review by dmonark — APPROVED** [review]

File: types/services/database.go [hunk 3]
```diff
@@ -92,7 +92,7 @@ func FindOne(id int64) (*Service, error) {
 		return nil, errors.Missing(&Service{}, id)
 	}
 	service := &Service{}
-	db.First(&service, id)
+	db.Preload("Incidents").First(&service, id)
```
Comment (akash-sachan) [inline types/services/database.go]: Can we use constants?

File: handlers/incident.go [hunk 4]
```diff
@@ -83,6 +150,29 @@ func apiCreateIncidentHandler(w http.ResponseWriter, r *http.Request) {
 	sendJsonAction(incident, "create", w, r)
 }
 
+func apiUpdateIncidentHandler(w http.ResponseWriter, r *http.Request) {
+	incident, _, err := findIncident(r)
```
Comment (akash-sachan) [inline handlers/incident.go:158]: A bit hard to understand how different `incident` and `incidentVar` are

File: handlers/incident.go [hunk 5]
```diff
@@ -83,6 +150,29 @@ func apiCreateIncidentHandler(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	incident.Title = incidentVar.Title
+	incident.Description = incidentVar.Description
+	if err := incident.Update(); err != nil {
+		sendErrorJson(err, w, r)
+		return
+	}
+
+	sendJsonAction(incident, "update", w, r)
```
Comment (akash-sachan) [inline handlers/incident.go:177]: use constants

File: handlers/incident.go [hunk 6]
```diff
@@ -33,6 +36,70 @@ func apiServiceIncidentsHandler(w http.ResponseWriter, r *http.Request) {
...
+		return true
+	}
+	if !(incidentUpdates[len(incidentUpdates)-1].Type == resolved && timeDiff(incidentUpdates[len(incidentUpdates)-1]) > incidentsTimeoutInMinutes) {
+		return true
+	}
+
+	return false
+}
+
+func timeDiff(update *incidents.IncidentUpdate) float64 {
```
Comment (akash-sachan) [inline handlers/incident.go]: Can we rename it to getTimeDiff?

File: handlers/incident.go [hunk 7]
```diff
@@ -33,6 +36,70 @@ func apiServiceIncidentsHandler(w http.ResponseWriter, r *http.Request) {
 	returnJson(service.Incidents, w, r)
 }
 
+func apiServiceIncidentsHandlerActive(w http.ResponseWriter, r *http.Request) {
```
Comment (akash-sachan) [inline handlers/incident.go]: Is `apiServiceIncidentsHandlerActive` as per the naming convention in this repo.
It is not very descriptive.

File: handlers/incident.go [hunk 8]
```diff
@@ -33,6 +36,70 @@ func apiServiceIncidentsHandler(w http.ResponseWriter, r *http.Request) {
...
+	return visibleIncidents
+}
+
+func reverse(incidents []*incidents.IncidentUpdate) {
+	for i, j := 0, len(incidents)-1; i < j; i, j = i+1, j-1 {
+		incidents[i], incidents[j] = incidents[j], incidents[i]
+	}
+}
+
+func visibilityCheck(incident *incidents.Incident) bool {
```
Comment (akash-sachan) [inline handlers/incident.go]: can be renamed to checkVisibility

Comment (akash-sachan) [reply handlers/incident.go]: or isVisible

Comment (akash-sachan) [reply handlers/incident.go]: Can be made a method of the incident pointer also

File: handlers/incident.go [hunk 9]
```diff
@@ -33,6 +36,70 @@ func apiServiceIncidentsHandler(w http.ResponseWriter, r *http.Request) {
...
+	incidentUpdates := incident.Updates
+	log.Infof(fmt.Sprintf("Latest Incident Update: %v, Time Diff: %v ", &incidentUpdates[len(incidentUpdates)-1], timeDiff(incidentUpdates[len(incidentUpdates)-1])))
+	if len(incidentUpdates) == 0 {
```
Comment (akash-sachan) [inline handlers/incident.go]: This can be put in a function called eg. `hasZeroUpdates`

File: handlers/incident.go [hunk 10]
```diff
@@ -33,6 +36,70 @@ func apiServiceIncidentsHandler(w http.ResponseWriter, r *http.Request) {
...
+		return true
+	}
+	if !(incidentUpdates[len(incidentUpdates)-1].Type == resolved && timeDiff(incidentUpdates[len(incidentUpdates)-1]) > incidentsTimeoutInMinutes) {
```
Comment (akash-sachan) [inline handlers/incident.go]: This condition is a bit complex. Can be put in another function`checkResolutionTime` or something

File: types/services/database.go (hunk 3 above)
Comment (kRhythm53) [reply types/services/database.go]: DONE

File: handlers/incident.go (hunk 4 above)
Comment (kRhythm53) [reply handlers/incident.go:158]: changed the name to updatedIncident

File: handlers/incident.go (hunk 6 above)
Comment (kRhythm53) [reply handlers/incident.go]: Done

File: handlers/incident.go (hunk 9 above)
Comment (kRhythm53) [reply handlers/incident.go]: updated

File: handlers/incident.go (hunk 10 above)
Comment (kRhythm53) [reply handlers/incident.go]: updated

File: handlers/incident.go (hunk 8 above)
Comment (kRhythm53) [reply handlers/incident.go]: Done

**Pull request merged**